task/src/global_registry.rs

```diff
@@ -84,9 +87,29 @@ impl GlobalRegistry {
 
     /// update the cache
     async fn update_cache(&mut self) {
+        // NOTE: this can be done much more cleverly
+        // avoid one intersection by comparing max keys (constant time op vs O(n + m))
+        // and debatable how often the other op needs to be run
+        // probably much much less often
         let list = self.state_list.read().await;
-        for i in self.state_cache.len()..list.len() {
-            self.state_cache.push(list[i].clone());
+        let list_keys: BTreeSet<usize> = list.iter().map(|(k, _v)| k).cloned().collect();
+        let cache_keys: BTreeSet<usize> =
+            self.state_cache.iter().map(|(k, _v)| k).cloned().collect();
+        // bleh not as efficient
+        let missing_key_list = list_keys.difference(&cache_keys);
+        let expired_key_list = cache_keys.difference(&list_keys);
```

Comment:
Aren't these two `difference`s supposed to be the same?

Reply:
I think we need two different cases here. One for keys in the state_list that need to be inserted into the cache, and one for the keys in the cache that have been shut down and removed from the state_list.

Reply:
Ahh makes sense! `a.difference(b)` isn't the same as `b.difference(a)`.